### PR TITLE
(cloud-merge) Fix filecache coredump because the fileWriterMapKey is not unique

### PR DESCRIPTION
**Pull request by Lchangliang**
## Proposed changes

Followup #30039

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Lchangliang** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

File: be/src/io/cache/fs_file_cache_storage.h [hunk 1]
```diff
@@ -86,12 +86,19 @@ class FSFileCacheStorage : public FileCacheStorage {
 
     void load_cache_info_into_memory(BlockFileCache* _mgr) const;
 
+    using FileWriterMapKey = std::pair<UInt128Wrapper, size_t>;
+    struct FileWriterMapKeyHash {
+        std::size_t operator()(const FileWriterMapKey& w) const {
+            return UInt128Hash()(w.first.value_) + w.second;
```
Comment (gavinchou) [inline be/src/io/cache/fs_file_cache_storage.h:92]: This hash function cannot guarantee uniqueness and it is not robust
1. narrow down 128bit to 64bit increase the possibility of collision 
2. the `+` result may overflow and wrap around
3. the output distribution of the hash function is not uniform

try concat 128bit input hash and the 64bit size as a byte string and hash it.
e.g.
```
char* p1 = (char*) &(w.first.value_);
char* p2 = (char*) &(w.second);
char buf[16] {p1[0:15] .. p2[0:7]};
std::string_view str(buf, 24);
return std::hash(str);
```

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by Lchangliang** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**